client/state/experiments/reducer.ts

```diff
@@ -0,0 +1,68 @@
+/**
+ * External Dependencies
+ */
+import { Action } from 'redux';
+
+/**
+ * Internal Dependencies
+ */
+import { EXPERIMENT_FETCH, EXPERIMENT_ASSIGN, CURRENT_USER_RECEIVE } from 'state/action-types';
+import { ExperimentState, ExperimentAssign } from 'state/experiments/types';
+import { tracksAnonymousUserId } from 'lib/analytics/ad-tracking';
+
+/**
+ * Attempt to get the anon id for the user, if set
+ */
+export function getAnonIdFromCookie(): string | null {
+	const id = tracksAnonymousUserId();
+
+	return id == null || id === '' ? null : id;
+}
+
+const appStartedAt = Date.now();
+
+const resetState: ( anonId: string | null ) => ExperimentState = anonId => ( {
+	anonId,
+	isLoading: true,
+	nextRefresh: appStartedAt,
+	variations: null,
+} );
+
+export default function reducer( state: ExperimentState = resetState( null ), action: Action ) {
+	switch ( action.type ) {
+		/**
+		 * Store the anon id when we first load the application
+		 */
+		case '@@INIT':
+			state.anonId = getAnonIdFromCookie();
+			break;
+
+		/**
+		 * Store the user's assignment from the API
+		 */
+		case EXPERIMENT_ASSIGN:
+			state = {
+				...state,
+				isLoading: false,
+				variations: ( action as ExperimentAssign ).variations,
+				nextRefresh: ( action as ExperimentAssign ).nextRefresh,
+			};
+			break;
+
+		/**
+		 * Start retrieving the user's assignment from the API
+		 */
+		case EXPERIMENT_FETCH:
+			state.isLoading = true;
```

Comment:
The reducer must not modify the `state` argument. It can only return a new `state`. In this statement, you are modifying the old state. Then any code that compares the old and new state will fail to notice the change:
```js
const oldState = store.getState();
console.log( oldState.isLoading ); // logs 'false'
store.dispatch( { type: EXPERIMENT_FETCH } );
const newState = store.getState();
console.log( oldState.isLoading, newState.isLoading ); // both are true!
```
It's better to return a new state based on the old one:
```js
return {
  ...state,
  isLoading: true
}
```
Generally, I'd discourage doing any `state = ...` assignment in a reducer. Just return the new state right away.